File: BreathFirstSearch/네트워크.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>
using namespace std;

int current, nxt;
bool visited[201];

int solution(int n, vector<vector<int>> computers)
{
    int answer = 0;

    queue<int> q;
    for (int j = 0; j < computers.size(); ++j)
    {
        if (!visited[j])
        {
            answer++;
            q.push(j);
            visited[j] = true;
            while (!q.empty())
            {
                current = q.front();
                q.pop();
                for (int i = 0; i < computers[current].size(); ++i)
                {
                    if (!visited[i] && computers[current][i])
                    {
                        visited[i] = true;
                        q.push(i);
                    }
                }
            }
        }
    }
    return answer;
}

//// FOR TEST
// vector<vector<int>> v, v2;
// int main()
// {
//     v.resize(3);
//     v2.resize(3);
//     for (int i = 0; i < 3; ++i)
//     {
//         v[i].push_back(i);
//         v2[i].push_back(i);
//     }
//     v[0].push_back(1);
//     v[1].push_back(0);

//     v2[0].push_back(1);
//     v2[1].push_back(0);
//     v2[1].push_back(2);
//     v2[2].push_back(1);
    
//     // cout << solution(3, v) << endl;
//     cout << solution(3, v2) << endl;
// }

File: DepthFirstSearch/네트워크.py
import sys
sys.setrecursionlimit(50000)

visited = [False for i in range(201)]
def dfs(n, computers, x):
    for i in range(n):
        if i!=x and computers[x][i] and not visited[i]:
            visited[i] = True
            dfs(n, computers, i)

def solution(n, computers):
    answer = 0
    for i in range(n):
        for j in range(n):
            if computers[i][j] and not visited[j]:
                visited[j] = True
                dfs(n, computers, j)
                answer+=1
    return answer

print(solution(3, [[1,1,0],[1,1,0],[0,0,1]]))
print(solution(3, [[1,1,0],[1,1,1],[0,1,1]]))